### PR TITLE
openssl: Also bootstrap CA bundle using System keychain

### LINKED ISSUE #5291: The description for openssl installation is misleading
### Bug reports:

The caveat shows:
A CA file has been bootstrapped using certificates from the system
    keychain. To add additional certificates, place .pem files in ...

However, the only keychain used is the keychain /System/Library/Keychains/SystemRootCertificates.keychain which is NOT the system keychain. I myself interpret the system keychain as the "system" keychain, not the "system roots" keychain ...
### Formula Requests:

I'd suggest simply adapting the caveat to say system roots keychain instead of the system keychain. I'd like feedback on this before opening a PR.

### PR DESCRIPTION
**Pull request by Firefishy**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
Add `/Library/Keychains/System.keychain` to the keychains used for for seeding the cert.pem bundle.

`System.keychain` is the authorative keychain for organization installed CAs.

`SystemRootCertificates.keychain` is only for default Apple installed CAs.

Fixes: https://github.com/Homebrew/homebrew-core/issues/5291
and likely many others where certificates were unexpectedly untrusted by homebrew installed apps but supported in browsers.

### CODE REVIEW

**Comment by Firefishy** [comment]
https://github.com/raggi/openssl-osx-ca is a tap for adding a service to keep keychain CAs and CA bundle `cert.pem` in sync. Its export script https://github.com/raggi/openssl-osx-ca/blob/master/osx-ca-certs.rb is a cleaner implementation than that used in in the `macos_post_instal`l section

**Comment by Firefishy** [comment]
gnutls likely also needs to be updated: https://github.com/Homebrew/homebrew-core/blob/master/Formula/gnutls.rb#L68-L91

**Comment by Firefishy** [comment]
CI failure look to me like a false positive.

**Comment by jonchang** [comment]
We intentionally only use the Apple-suppled certificates when bootstrapping the openssl root certificate store. See https://github.com/Homebrew/homebrew-core/pull/1072 for our reasoning. Unless there is now a way to address that security issue we won't be accepting this change.

**Comment by Firefishy** [comment]
> We intentionally only use the Apple-suppled certificates when bootstrapping the openssl root certificate store. See #1072 for our reasoning. Unless there is now a way to address that security issue we won't be accepting this change.

Interesting. Yes there is a method to verify that a certificate is trusted in Keychain: `/usr/bin/security verify-cert -q -l -L -R offline -c ca-cert-file-to-verify.pem`. Here is an example implementation: https://github.com/raggi/openssl-osx-ca/blob/master/osx-ca-certs.rb#L11-L29

I am happy to re-work the PR to add the verification step.

**Comment by bayandin** [comment]
The new version of `openssl@1.1` 1.1.1j has been released: https://github.com/Homebrew/homebrew-core/pull/71329.
If we're going to accept these changes we need either add it to that PR or rebase this one when https://github.com/Homebrew/homebrew-core/pull/71329 is merged.

**Comment by carlocab** [comment]
I've cancelled the tests. There's no reason to have two OpenSSL CI jobs running in parallel.

Please wait until you're able to rebase against #71329 (once that is merged) before pushing more changes here.

**Comment by Firefishy** [comment]
Added functionality to address issues from #1072

@carlocab Feel free to cancel CI jobs if desired.

**Comment by Firefishy** [comment]
Rebased against #71329 changes.

**Comment by Firefishy** [comment]
This is ready for review.

**Comment by fxcoudert** [comment]
On my system, this is outputting:

```
$ time brew postinstall openssl
==> Postinstalling openssl@1.1
==> Regenerating CA certificate bundle from keychain, this may take a while
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
SecCertificateCreateFromData: Unknown format in import.
brew postinstall openssl  6.04s user 3.77s system 134% cpu 7.301 total
```

**Comment by Firefishy** [comment]
> ```
> $ time brew postinstall openssl
> ==> Postinstalling openssl@1.1
> ==> Regenerating CA certificate bundle from keychain, this may take a while
> Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
> Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
> Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
> Cert Verify Result: CSSMERR_TP_NOT_TRUSTED
> Cert Verify Result: CSSMERR_TP_NOT_TRUSTED 

These are normal messages from `/usr/bin/security verify-cert`, but maybe should be silenced. You have certificates which are not set to TRUSTED in likely system keychain. They should NOT have been added to the `cert.pem`  

> ```
> SecCertificateCreateFromData: Unknown format in import.


Unsure what caused this error, but likely safely skipped that cert and continued the generation of `cert.pem`

**Comment by Firefishy** [comment]
I use the following command to check the contents of `cert.pem`:
```
openssl crl2pkcs7 -nocrl -certfile /usr/local/etc/openssl@1.1/cert.pem | openssl pkcs7 -print_certs -text -noout |grep -F Subject:
```

**Comment by Firefishy** [comment]
@fxcoudert Would you like me to silence those messages output by `/usr/bin/security verify-cert`? I don't see them on my setup. I silence stdout, but may be stderr messages. I've seen the issue on older MacOS releases.

**Comment by fxcoudert** [comment]
If that output is expected, yes we need to silence it.

**Comment by Firefishy** [comment]
> If that output is expected, yes we need to silence it.

Done.

**Comment by Firefishy** [comment]
@fxcoudert Silenced the expected output.
Also minor cleanups.

The build failures look like false positives.

**Comment by Firefishy** [comment]
Anything I can do to progress this?

**Comment by fxcoudert** [comment]
One of the original arguments against this was the time it took at installation. I'm not seeing this, as it's relatively fast on my real-life laptop (6 seconds). Is the approach you've chosen different, explaining why this is much shorter than the 5 or so minutes reported in a previous attempt?

@Homebrew/core depending on the answer above, I'm inclined to accept the PR. Does anyone have objections?

**Comment by Firefishy** [comment]
> One of the original arguments against this was the time it took at installation. I'm not seeing this, as it's relatively fast on my real-life laptop (6 seconds). Is the approach you've chosen different, explaining why this is much shorter than the 5 or so minutes reported in a previous attempt?
> 

I guess it is faster because we only add CA certs from the 2 keychains (Apple built-in & system) and do not include User keychain which can sometimes contain many more certificates. We also only run `/usr/bin/security verify-cert` on certificates that haven't already expired. `/usr/bin/security verify-cert` also runs in offline mode (using existing revocation cache) which would also help.

File: Formula/openssl@1.1.rb
Comment (MikeMcQuaid) [inline Formula/openssl@1.1.rb:120]:
```suggestion
    ohai "Regenerating CA certificate bundle from keychain, this may take a while..."
```

File: Formula/openssl@1.1.rb
Comment (MikeMcQuaid) [inline Formula/openssl@1.1.rb:135]:
```suggestion
      IO.popen("#{bin}/openssl x509 -inform pem -checkend 0 -noout 2>&1 >/dev/null", "w") do |openssl_io|
```

File: Formula/openssl@1.1.rb
Comment (MikeMcQuaid) [inline Formula/openssl@1.1.rb:160]:
```suggestion
    trusted_certs = begin
      tmpfile = Tempfile.new

        valid_certs.select do |cert|
          tmpfile.rewind
          tmpfile.write cert
          tmpfile.truncate cert.size
          tmpfile.flush
          cmd = "/usr/bin/security verify-cert -l -L -R offline -c #{tmpfile.path} 2>&1 >/dev/null"
          IO.popen(cmd)

          $CHILD_STATUS.success?
        end
      ensure
        tmpfile&.close!
      end
```

File: Formula/openssl@1.1.rb
Comment (MikeMcQuaid) [inline Formula/openssl@1.1.rb:154]:
```suggestion
          IO.popen("/usr/bin/security verify-cert -l -L -R offline -c #{tmpfile.path} &>/dev/null")
```

**Review by MikeMcQuaid — DISMISSED** [review]
I'm also 👍🏻 on this being merged given the explanations here. Nice work @Firefishy!

File: Formula/openssl@1.1.rb
Comment (Firefishy) [reply Formula/openssl@1.1.rb:135]: I was going to do it this way, but it then fails the line length test

Comment (MikeMcQuaid) [reply Formula/openssl@1.1.rb]:
```suggestion
      IO.popen("#{bin}/openssl x509 -inform pem -checkend 0 -noout &>/dev/null", "w") do |openssl_io|
```

may work, if not, leave it as is (but use `&>`)

Comment (Firefishy) [reply Formula/openssl@1.1.rb]: Done. 👍  Learnt some new bash shorthand

**Comment by sjackman** [comment]
@woodruffw Any thoughts on this PR?

**Comment by Firefishy** [comment]
@raggi Hope you are also ok with this PR being merged? As you know the validation portion strongly inspired by your code https://github.com/raggi/openssl-osx-ca/blob/master/osx-ca-certs.rb

**Comment by Firefishy** [comment]
Rebased to pull in the fixes for https://github.com/Homebrew/homebrew-core/pull/72928 and  https://github.com/Homebrew/homebrew-core/pull/73166

**Comment by woodruffw** [comment]
> @woodruffw Any thoughts on this PR?

I took a quick scan, and it looks reasonable to me. I think most users expect `System.keychain` to be included in the CA, and that including it doesn't significantly change the user's security posture (you already have to be an administrator to modify `System.keychain`, the Homebrewed CA bundle can be modified in other ways with fewer permissions, and anybody being snooped because of corporate policies has already bitten the bullet with their non-Homebrewed bundles in other software).

So, LGTM 🙂

**Review by SMillerDev — APPROVED** [review]
Thanks @Firefishy  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Comment by reegnz** [comment]
Is there a way to re-run the macos_post_install script again to rebuild the cert.pem from the macos keychain?
I am still using https://github.com/raggi/openssl-osx-ca because of that purpose, but the code is already in this formula, and I was wondering if I could just re-run that macos_post_install script when the System Keychain changes.

**Comment by bayandin** [comment]
> Is there a way to re-run the macos_post_install script again to rebuild the cert.pem from the macos system keychain?

`$ brew postinstall openssl@1.1`

**Comment by m-urban** [comment]
Hi @Firefishy, @SMillerDev,

I would have loved to leave this comment on #71282 but that's closed already. Anyways:

I am on macOS 10.14.6 Mojave and I am afraid that adding the `System.keychain` to the CA bundle reproducibly breaks my gnutls installation.
As of the recent change, e.g. `gnutls-cli -d 1 google.com` keeps failing with the `System.keychain` included in `gnutls.rb`'s `macos_post_install` action. Once I remove the line from the script and `brew postinstall gnutls`, the `gnutls-cli` command goes through again.

Does gnutls with the `System.keychain` included work correctly for you?

**Comment by Firefishy** [comment]
Yes. I did test gnutls-cli and it worked for me. There were a few CA certificate warning because gnutils is stricter than what `/usr/bin/security verify-cert` is.

I will check again later.

**Comment by SMillerDev** [comment]
> I am on macOS 10.14.6 Mojave and I am afraid that adding the System.keychain to the CA bundle reproducibly breaks my gnutls installation.

We've had people's antivirus block this. Try if that's the cause

**Comment by m-urban** [comment]
Thanks for getting back to me so quickly! I do not have an anti-virus program installed (but trust Little Snitch as a firewall solution).

Other than than that, I checked the System keychain in Keychain Access and only found three certificates in there:
* Apple Worldwide Developer Relations Certification Authority (shown as valid)
* com.apple.kerberos.kdc (shown as self-signed, not verified by a root CA)
* com.apple.systemdefault (shown as self-signed, not verified by a root CA)

Is it possible that the latter two certificates break the *.pem-bundle?

**Comment by SMillerDev** [comment]
Maybe check if `/usr/bin/security verify-cert` connects to the internet (and is blocked by Little Snitch)

**Comment by m-urban** [comment]
I am not quite sure what to do with your command but looking into the macos_post_install script, I tried to replicate what the script does and ran `security verify-cert -l -L -R offline -c /usr/local/etc/gnutls/cert.pem`

Results:
* Without the System.keychain, the command returns `...certificate verification successful.`
* With the System.keychain included, the command returns `Cert Verify Result: CSSMERR_TP_NOT_TRUSTED`

Digging deeper, I played around with the trust settings in Keychain Access and found the culprit to be the certificate _com.apple.systemdefault_. When setting the trust setting for this particulate cert to _Trust Always_ instead of _System Default_ (not sure of the English localizations, German user here) and run the postinstall action again, the `security verify-cert…` command goes through successfully.
The _com.apple.systemdefault_ cert seems to be a, well…, system default one, that is part of probably every macOS installation, if my interpretation of a quick Google search is correct.

However, none of that seems to have to do with my failing `gnutls-cli -d 1 google.com`. That still only goes through if I entirely remove the System.keychain from the script and re-build the cert bundle.

**Comment by Firefishy** [comment]
Can you post the full output of `gnutls-cli -d 1 google.com`, feel free to obscure some of the `There was a non-CA certificate in the trusted list` warnings.

Mine:
```
$ gnutls-cli -d 1 google.com
|<1>| There was a non-CA certificate in the trusted list: CN=com.apple.systemdefault,O=System Identity.
|<1>| There was a non-CA certificate in the trusted list: CN=com.apple.kerberos.kdc,O=System Identity.
|<1>| There was a non-CA certificate in the trusted list: CN=*REMOVED RELATED TO EMPLOYER*
|<1>| There was a non-CA certificate in the trusted list: OU=*REMOVED RELATED TO EMPLOYER*
|<1>| There was a non-CA certificate in the trusted list: CN=*REMOVED RELATED TO EMPLOYER*
|<1>| There was a non-CA certificate in the trusted list: OU=*REMOVED RELATED TO MACHINE CERT*
|<1>| There was a non-CA certificate in the trusted list: CN=*REMOVED RELATED TO EMPLOYER*
|<1>| There was a non-CA certificate in the trusted list: CN=*REMOVED RELATED TO EMPLOYER*
Processed 175 CA certificate(s).
Resolving 'google.com:443'...
Connecting to '142.250.187.238:443'...
- Certificate type: X.509
- Got a certificate list of 2 certificates.
- Certificate[0] info:
 - subject `CN=*.google.com,O=Google LLC,L=Mountain View,ST=California,C=US', issuer `CN=GTS CA 1O1,O=Google Trust Services,C=US', serial 0x00bff10d86136f613d0300000000cc17de, EC/ECDSA key 256 bits, signed using RSA-SHA256, activated `2021-05-10 01:34:19 UTC', expires `2021-08-02 01:34:18 UTC', pin-sha256="rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo="
	Public Key ID:
		sha1:4bd0097b78a741197015ac93582358cc24ca0df2
		sha256:aeaaa0d83f351820b0b9f65ba188565ff7e33090b46e1b8a5fc4b20181c6ad1a
	Public Key PIN:
		pin-sha256:rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo=

- Certificate[1] info:
 - subject `CN=GTS CA 1O1,O=Google Trust Services,C=US', issuer `CN=GlobalSign,O=GlobalSign,OU=GlobalSign Root CA - R2', serial 0x01e3b49aa18d8aa981256950b8, RSA key 2048 bits, signed using RSA-SHA256, activated `2017-06-15 00:00:42 UTC', expires `2021-12-15 00:00:42 UTC', pin-sha256="YZPgTZ+woNCCCIW3LH2CxQeLzB/1m42QcCTBSdgayjs="
- Status: The certificate is trusted.
- Description: (TLS1.3-X.509)-(ECDHE-X25519)-(ECDSA-SECP256R1-SHA256)-(AES-256-GCM)
- Options:
- Handshake was completed

- Simple Client Mode:
```

**Comment by m-urban** [comment]
Using the cert bundle created without the System.keychain, the output looks identical to yours:
```
$ gnutls-cli -d 1 google.com
Processed 165 CA certificate(s).
Resolving 'google.com:443'...
Connecting to '172.217.22.238:443'...
- Certificate type: X.509
- Got a certificate list of 2 certificates.
- Certificate[0] info:
 - subject `CN=*.google.com,O=Google LLC,L=Mountain View,ST=California,C=US', issuer `CN=GTS CA 1O1,O=Google Trust Services,C=US', serial 0x00bff10d86136f613d0300000000cc17de, EC/ECDSA key 256 bits, signiert mit RSA-SHA256, activated `2021-05-10 01:34:19 UTC', expires `2021-08-02 01:34:18 UTC', pin-sha256="rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo="
	Public Key ID:
		sha1:4bd0097b78a741197015ac93582358cc24ca0df2
		sha256:aeaaa0d83f351820b0b9f65ba188565ff7e33090b46e1b8a5fc4b20181c6ad1a
	Public Key PIN:
		pin-sha256:rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo=

- Certificate[1] info:
 - subject `CN=GTS CA 1O1,O=Google Trust Services,C=US', issuer `CN=GlobalSign,O=GlobalSign,OU=GlobalSign Root CA - R2', serial 0x01e3b49aa18d8aa981256950b8, RSA key 2048 bits, signiert mit RSA-SHA256, activated `2017-06-15 00:00:42 UTC', expires `2021-12-15 00:00:42 UTC', pin-sha256="YZPgTZ+woNCCCIW3LH2CxQeLzB/1m42QcCTBSdgayjs="
- Status: The certificate is trusted.
- Description: (TLS1.3-X.509)-(ECDHE-X25519)-(ECDSA-SECP256R1-SHA256)-(AES-256-GCM)
- Options:
- Handshake was completed

- Simple Client Mode:
```

Using the cert bundle including the System.keychain, the command fails with:
```
$ gnutls-cli -d 1 google.com
Processed 0 CA certificate(s).
Resolving 'google.com:443'...
Connecting to '172.217.22.238:443'...
- Certificate type: X.509
- Got a certificate list of 2 certificates.
- Certificate[0] info:
 - subject `CN=*.google.com,O=Google LLC,L=Mountain View,ST=California,C=US', issuer `CN=GTS CA 1O1,O=Google Trust Services,C=US', serial 0x00bff10d86136f613d0300000000cc17de, EC/ECDSA key 256 bits, signiert mit RSA-SHA256, activated `2021-05-10 01:34:19 UTC', expires `2021-08-02 01:34:18 UTC', pin-sha256="rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo="
	Public Key ID:
		sha1:4bd0097b78a741197015ac93582358cc24ca0df2
		sha256:aeaaa0d83f351820b0b9f65ba188565ff7e33090b46e1b8a5fc4b20181c6ad1a
	Public Key PIN:
		pin-sha256:rqqg2D81GCCwufZboYhWX/fjMJC0bhuKX8SyAYHGrRo=

- Certificate[1] info:
 - subject `CN=GTS CA 1O1,O=Google Trust Services,C=US', issuer `CN=GlobalSign,O=GlobalSign,OU=GlobalSign Root CA - R2', serial 0x01e3b49aa18d8aa981256950b8, RSA key 2048 bits, signiert mit RSA-SHA256, activated `2017-06-15 00:00:42 UTC', expires `2021-12-15 00:00:42 UTC', pin-sha256="YZPgTZ+woNCCCIW3LH2CxQeLzB/1m42QcCTBSdgayjs="
- Status: The certificate is NOT trusted. The certificate issuer is unknown.
*** PKI verification of server certificate failed...
*** Fatal error: Fehler im Zertifikat. [← m-urban: translation "error in the certificate"]
```

**Comment by Firefishy** [comment]
Issue is the bundle has no CA certificates in the second example, specifically `Processed 0 CA certificate(s).`

**Comment by m-urban** [comment]
Ok, so it seems that the post-install script generates an invalid bundle (= `gnutls-cli` comes back with `Processed 0 CA certificate(s)`). I verified that the generated bundle is invalid in the following cases:
* When the bundle exlusively only contains the System.keychain
* When the bundle contains SystemRootCertificates.keychain and System.keychain
* When the bundle contains System.keychain and SystemRootCertificates.keychain (reverse order)

**Comment by MichaelEischer** [comment]
I can confirm that it's one of the certificates in the bundle that breaks the CA file. For me, after removing the second certificate in `cert.pem` and storing the certificate in `broken.crt` I get the following output:
```
$ gnutls-certtool --infile=broken.crt --verify
Note that no verification profile was selected. In the future the medium profile will be enabled by default.
Use --verify-profile low to apply the default verification of NORMAL priority string.
|<1>| There was a non-CA certificate in the trusted list: CN=com.apple.systemdefault,O=System Identity.
Loaded system trust (164 CAs available)
error parsing CRTs: Duplicate extension in X.509 certificate.
```

If I add that certificate back to the ca list, then `gnutls-certtool`  silently fails to load any cert from the ca list: `Loaded system trust (0 CAs available)`.

**Comment by sCreami** [comment]
I'd like to add to the discussion that bootstrapping /Library/Keychains/System.keychain also broke my gnutls config. I confirm the generated bundle is invalid. It took me half a day to figure out why I couldn't perform any TLS handshake.

**Comment by Firefishy** [comment]
> I can confirm that it's one of the certificates in the bundle that breaks the CA file. For me, after removing the second certificate in `cert.pem` and storing the certificate in `broken.crt` I get the following output:
> 

Cloud you see if `broken.crt` can be read by openssl? With command: `openssl crl2pkcs7 -nocrl -certfile broken.crt | openssl pkcs7 -print_certs -text -noout`

Th output would be helpful.

**Comment by MichaelEischer** [comment]
@Firefishy That command gives me the following output (I've removed a few lines and marked these as "lines removed" as I'm not sure whether some of it is supposed to be secret or not):
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 725025595 (0x2b37033b)
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=com.apple.kerberos.kdc, O=System Identity
        Validity
            Not Before: Jan 25 13:07:46 2017 GMT
            Not After : Jan 20 13:07:46 2037 GMT
        Subject: CN=com.apple.kerberos.kdc, O=System Identity
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:d3:65:00:57:e9:51:db:79:5a:07:61:bf:dc:61:
                    3c:00:05:92:dd:31:b0:99:e9:44:6e:3d:73:c8:5d:
                    f6:5c:b9:f0:c6:cd:41:b7:00:c3:69:c2:f0:7f:2e:
                    51:94:19:f7:21:08:25:20:84:ad:c2:30:c4:7c:6e:
                    [13 lines removed]
                    53:e7
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: 
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Extended Key Usage: 
                1.3.6.1.5.2.3.5
    Signature Algorithm: sha256WithRSAEncryption
         4c:8c:33:13:af:2c:d1:d0:62:48:8b:7f:dd:75:1f:97:d5:12:
         b4:08:4d:aa:04:e4:cd:bb:32:82:dd:f2:45:2c:5e:a5:34:9b:
         8c:6e:78:b5:3b:11:c2:bf:24:48:1a:17:13:52:32:36:3f:75:
         9d:f7:29:59:e7:6e:84:12:5d:ae:19:fa:46:cc:90:4d:59:6f:
         [10 lines removed]
         0d:57:50:47
```

**Pull request closed without merging**
